File: frontend/src/lib/reports/cashflow.ts
import type { BalanceQuantity } from '$lib/api/ledger';
import type { CashflowReportResponse } from '$lib/api/reports';

/**
 * One table row: a single calendar bucket in a single commodity.
 *
 * The response carries seven independent per-commodity lists per bucket. A
 * commodity may appear in some and not others — a month with only a card
 * payment has a transfer and a net movement but no inflow — so the rows are
 * built from the union of commodities present, with an absent measure read as
 * zero rather than as a missing row.
 */
export type CashflowRow = {
  key: string;
  startDate: string;
  endDate: string;
  commodityID: number;
  /**
   * The deepest scale any of this row's measures carries, for anything that
   * needs one number for the row. Never use it to render a measure: read that
   * measure's own entry in `scales`.
   */
  quantityScale: number;
  /**
   * Each measure's own scale.
   *
   * The server writes every measure of one commodity at one scale, but a
   * coefficient must be paired with the scale it was produced at rather than
   * with a scale borrowed from a sibling measure: the two differ by a power of
   * ten per digit, so borrowing renders 100 as 1.00.
   */
  scales: Record<Measure, number>;
  inflow: string;
  outflow: string;
  operatingNet: string;
  transferIn: string;
  transferOut: string;
  transferNet: string;
  netMovement: string;
  /**
   * True for the reporting-currency restatement of a bucket. Its three net
   * measures are derived from its own converted parts, so the identities hold
   * within the row exactly as they do in a commodity row.
   */
  converted: boolean;
};

export type Measure =
  | 'inflow'
  | 'outflow'
  | 'operatingNet'
  | 'transferIn'
  | 'transferOut'
  | 'transferNet'
  | 'netMovement';

const MEASURES: Array<[Measure, keyof CashflowReportResponse['buckets'][number]]> = [
  ['inflow', 'inflow'],
  ['outflow', 'outflow'],
  ['operatingNet', 'operating_net'],
  ['transferIn', 'transfer_in'],
  ['transferOut', 'transfer_out'],
  ['transferNet', 'transfer_net'],
  ['netMovement', 'net_movement']
];

function byCommodity(quantities: BalanceQuantity[] | undefined): Map<number, BalanceQuantity> {
  const map = new Map<number, BalanceQuantity>();
  for (const quantity of quantities ?? []) {
    map.set(quantity.commodity_id, quantity);
  }
  return map;
}

export function cashflowRows(report: CashflowReportResponse): CashflowRow[] {
  return report.buckets.flatMap((bucket) => {
    const measures = new Map<Measure, Map<number, BalanceQuantity>>();
    for (const [measure, field] of MEASURES) {
      measures.set(measure, byCommodity(bucket[field] as BalanceQuantity[]));
    }

    const commodityIDs = new Set<number>();
    for (const quantities of measures.values()) {
      for (const commodityID of quantities.keys()) {
        commodityIDs.add(commodityID);
      }
    }

    const rows = [...commodityIDs]
      .sort((a, b) => a - b)
      .map((commodityID) => {
        const row: CashflowRow = {
          key: `${bucket.start_date}-${bucket.end_date}-${commodityID}`,
          startDate: bucket.start_date,
          endDate: bucket.end_date,
          commodityID,
          quantityScale: 0,
          scales: {
            inflow: 0,
            outflow: 0,
            operatingNet: 0,
            transferIn: 0,
            transferOut: 0,
            transferNet: 0,
            netMovement: 0
          },
          inflow: '0',
          outflow: '0',
          operatingNet: '0',
          transferIn: '0',
          transferOut: '0',
          transferNet: '0',
          netMovement: '0',
          converted: false
        };

        for (const [measure] of MEASURES) {
          const quantity = measures.get(measure)?.get(commodityID);
          if (!quantity) continue;
          row[measure] = quantity.quantity_value;
          row.scales[measure] = quantity.quantity_scale;
          row.quantityScale = Math.max(row.quantityScale, quantity.quantity_scale);
        }
        // An absent measure reads as zero, and zero is the same number at every
        // scale, so it takes the row's scale rather than a misleading 0.
        for (const [measure] of MEASURES) {
          if (!measures.get(measure)?.get(commodityID)) {
            row.scales[measure] = row.quantityScale;
          }
        }
        return row;
      });

    // A bucket with no commodity rows has nothing to restate. The backend
    // converts an empty bucket to an exact zero, which is true but is not
    // activity — emitting it would replace the "nothing moved" empty state with
    // a table of zeros.
    const restatement = rows.length === 0 ? null : convertedRow(bucket);
    // After the commodity rows: the restatement summarizes them, and reading it
    // first would suggest they break it down.
    return restatement ? [...rows, restatement] : rows;
  });
}

/**
 * A bucket's reporting-currency restatement, or null when the bucket did not
 * convert.
 *
 * Every measure has to be present for a row to be built. A partial row would
 * put a real inflow beside a zero outflow that only means "unknown", and the
 * net movement beside them would be neither.
 */
function convertedRow(bucket: CashflowReportResponse['buckets'][number]): CashflowRow | null {
  const parts: Array<[Measure, BalanceQuantity | undefined]> = [
    ['inflow', bucket.converted_inflow],
    ['outflow', bucket.converted_outflow],
    ['transferIn', bucket.converted_transfer_in],
    ['transferOut', bucket.converted_transfer_out],
    ['operatingNet', bucket.converted_operating_net],
    ['transferNet', bucket.converted_transfer_net],
    ['netMovement', bucket.converted_net_movement]
  ];
  if (parts.some(([, quantity]) => quantity === undefined)) return null;

  // Every converted figure is rounded once into the reporting currency's own
  // scale, so one scale covers the whole row.
  const scale = bucket.converted_net_movement!.quantity_scale;
  const value = (measure: Measure) =>
    parts.find(([name]) => name === measure)![1]!.quantity_value;

  return {
    key: `${bucket.start_date}-${bucket.end_date}-converted`,
    startDate: bucket.start_date,
    endDate: bucket.end_date,
    commodityID: bucket.converted_net_movement!.commodity_id,
    quantityScale: scale,
    scales: {
      inflow: scale,
      outflow: scale,
      operatingNet: scale,
      transferIn: scale,
      transferOut: scale,
      transferNet: scale,
      netMovement: scale
    },
    inflow: value('inflow'),
    outflow: value('outflow'),
    operatingNet: value('operatingNet'),
    transferIn: value('transferIn'),
    transferOut: value('transferOut'),
    transferNet: value('transferNet'),
    netMovement: value('netMovement'),
    converted: true
  };
}

export function heldRows(rows: CashflowRow[]): CashflowRow[] {
  return rows.filter((row) => !row.converted);
}

export function hasMultipleCommodities(rows: CashflowRow[]): boolean {
  return new Set(heldRows(rows).map((row) => row.commodityID)).size > 1;
}

/**
 * The converted restatements, but only when every bucket has one: a series
 * with a hole would chart a month of no movement that in fact did not convert.
 */
export function convertedSeries(rows: CashflowRow[]): CashflowRow[] {
  const buckets = new Set(rows.map((row) => `${row.startDate}/${row.endDate}`));
  const converted = rows.filter((row) => row.converted);
  return converted.length > 0 && converted.length === buckets.size ? converted : [];
}

/**
 * True when nothing moved anywhere in the range.
 *
 * A bucket whose movements all net to zero is still real activity, so the empty
 * state keys on there being no rows at all rather than on the totals reading
 * zero — telling a user "no activity" when their transfers cancelled out would
 * be wrong.
 */
export function isEmptyCashflow(rows: CashflowRow[]): boolean {
  return rows.length === 0;
}

/**
 * Which sides of the financing movement are worth naming under the net figure.
 *
 * A row with money moving one way only should not read "in 0.00 · out 750.00":
 * the zero half is noise, and in a dense table noise is what stops people
 * reading the column at all.
 */
export function transferDetail(row: CashflowRow): 'none' | 'in' | 'out' | 'both' {
  const incoming = row.transferIn !== '0';
  const outgoing = row.transferOut !== '0';
  if (incoming && outgoing) return 'both';
  if (incoming) return 'in';
  if (outgoing) return 'out';
  return 'none';
}

/**
 * The spending-report link that breaks down one bucket's inflow or outflow.
 *
 * Cashflow deliberately takes no category or payee filter: such a filter removes
 * counterpart postings from the basis, and `net_movement` would stop reconciling
 * to the cash balance change. The breakdown belongs to the spending report,
 * whose account filter is a counterpart filter over the same accounts — so the
 * two ask the same question, and this link is how you get from one to the other.
 *
 * The scope is passed as the *resolved* account ids the cashflow report actually
 * measured, so the drill-down inherits the exact set rather than re-deriving a
 * default that may have drifted since.
 */
export function cashflowDrillDownHref(
  row: CashflowRow,
  direction: 'in' | 'out',
  resolvedAccountIDs: number[],
  bucket: string
): string | undefined {
  // A restatement has no postings of its own. Its commodity is the reporting
  // currency, so the link would narrow the breakdown to whatever happens to be
  // held in that currency — a different set from the one the row sums. The
  // commodity rows it summarizes each keep their own link.
  if (row.converted) return undefined;

  const amount = direction === 'in' ? row.inflow : row.outflow;
  // Nothing moved in that direction, so there is nothing to break down.
  if (amount === '0') return undefined;

  const params = new URLSearchParams();
  params.set('view', 'spending');
  params.set('group_by', 'category');
  // An inflow is money from income counterparts, an outflow money to expense
  // counterparts — exactly the spending report's direction switch.
  params.set('mode', direction === 'in' ? 'income' : 'spending');
  params.set('start_date', row.startDate);
  params.set('end_date', row.endDate);
  // Spending ranks one range and ignores the bucket, but the reports screen
  // requires a valid one in the URL; carrying cashflow's own keeps a later
  // switch to net worth showing the same granularity.
  params.set('bucket', bucket);

  for (const accountID of resolvedAccountIDs) {
    params.append('account_id', String(accountID));
  }
  // The row's own commodity only. A cashflow row is per commodity, and widening
  // the breakdown to the others would answer a question the row did not ask.
  params.append('commodity_id', String(row.commodityID));

  return `/app/reports?${params.toString()}`;
}

File: backend/internal/app/cashflow.go
package app

import (
	"context"
	"fmt"
	"maps"
	"math/big"
	"slices"
	"strings"

	"rekenraam/backend/internal/db"
	"rekenraam/backend/internal/exact"
)

// defaultCashAccountKinds is the liquid-cash scope a cashflow report uses when
// the caller names no accounts (reports-plan.md, cashflow rule 1). It is a
// named default, not an invisible "all assets" shortcut: the response echoes
// the accounts it resolved to.
var defaultCashAccountKinds = map[string]bool{
	"cash":           true,
	"checking":       true,
	"savings":        true,
	"brokerage_cash": true,
}

type CashflowInput struct {
	// Reporting is optional; the exact per-commodity measures are unchanged
	// whether or not it is set.
	Reporting *ReportingCurrencyInput
	StartDate string
	EndDate   string
	Bucket    string
	Filters   ReportFilters
}

// CashflowBucket holds one calendar bucket's exact per-commodity totals.
//
// Every list is per commodity and unlike commodities are never combined. The
// identity net_movement = operating_net + transfer_net holds exactly in every
// commodity of every bucket, because journal entries balance per commodity:
// the counterparts of a cash posting account for precisely the movement it
// represents.
type CashflowBucket struct {
	StartDate string
	EndDate   string
	// Inflow and Outflow are positive magnitudes of money entering and leaving
	// the selected cash scope against income and expense counterparts.
	Inflow  []BalanceQuantity
	Outflow []BalanceQuantity
	// TransferIn and TransferOut are positive magnitudes of movement against
	// asset, liability, and equity counterparts outside the selected scope.
	// This never becomes income or spending.
	TransferIn  []BalanceQuantity
	TransferOut []BalanceQuantity
	// OperatingNet is Inflow - Outflow and excludes transfer/financing movement.
	OperatingNet []BalanceQuantity
	// TransferNet is TransferIn - TransferOut.
	TransferNet []BalanceQuantity
	// NetMovement is the signed sum of every selected-cash posting in the
	// bucket, so it reconciles exactly to the selected cash balance change
	// across the bucket's boundaries.
	NetMovement []BalanceQuantity
	// Converted holds the same seven measures in the reporting currency, each
	// summed from postings converted at their own entry date. Absent when a
	// posting in the bucket could not be converted. The three net measures are
	// derived from the converted parts so their identities hold exactly — see
	// cashflowTotals.
	ConvertedInflow       *BalanceQuantity
	ConvertedOutflow      *BalanceQuantity
	ConvertedTransferIn   *BalanceQuantity
	ConvertedTransferOut  *BalanceQuantity
	ConvertedTransferNet  *BalanceQuantity
	ConvertedOperatingNet *BalanceQuantity
	ConvertedNetMovement  *BalanceQuantity
}

type CashflowResult struct {
	StartDate           string
	EndDate             string
	Bucket              string
	Filters             ReportFiltersEcho
	CashScope           string
	CashAccountKinds    []string
	Buckets             []CashflowBucket
	ExcludedSystemRoles []string
	Valuation           *ValuationCoverage
}

// cashflowTotals accumulates one bucket's exact sums per commodity, and — when
// a reporting currency is asked for — the same measures converted.
//
// The converted net movement is **derived** from the converted parts rather
// than accumulated from the cash postings directly. Both are defensible; only
// one keeps R2's identity. Each posting is rounded to the reporting currency as
// it is converted, and rounding is not additive: converting a -100.00 cash
// posting can differ by a minor unit from converting its 60.00 and 40.00
// counterparts separately. Deriving makes net_movement = operating_net +
// transfer_net true by construction in the converted view, exactly as it is in
// the exact one, and the response names the method that produced it.
type cashflowTotals struct {
	inflow      map[int64]*exact.ScaledInt
	outflow     map[int64]*exact.ScaledInt
	transferIn  map[int64]*exact.ScaledInt
	transferOut map[int64]*exact.ScaledInt
	netMovement map[int64]*exact.ScaledInt

	rates                *RateTable
	convertedInflow      *ConvertedTotal
	convertedOutflow     *ConvertedTotal
	convertedTransferIn  *ConvertedTotal
	convertedTransferOut *ConvertedTotal
}

func newCashflowTotals(rates *RateTable) *cashflowTotals {
	totals := &cashflowTotals{
		inflow:      map[int64]*exact.ScaledInt{},
		outflow:     map[int64]*exact.ScaledInt{},
		transferIn:  map[int64]*exact.ScaledInt{},
		transferOut: map[int64]*exact.ScaledInt{},
		netMovement: map[int64]*exact.ScaledInt{},
	}
	if rates != nil {
		totals.rates = rates
		totals.convertedInflow = NewConvertedTotal(rates.Scale())
		totals.convertedOutflow = NewConvertedTotal(rates.Scale())
		totals.convertedTransferIn = NewConvertedTotal(rates.Scale())
		totals.convertedTransferOut = NewConvertedTotal(rates.Scale())
	}
	return totals
}

// addConverted values one counterpart posting at its own entry date — a flow,
// converted when it happened.
func (t *cashflowTotals) addConverted(target *ConvertedTotal, posting db.ReportCashflowPostingRecord, negate bool) {
	if t.rates == nil {
		return
	}
	value := posting.QuantityValue
	if negate {
		value = value.Negated()
	}
	converted, ok := t.rates.Convert(value, posting.QuantityScale, posting.CommodityID, posting.EntryDate)
	target.Add(converted, ok)
}

// Cashflow answers what changed the selected liquid cash, without treating a
// transfer as spending.
//
// The classification is exact rather than heuristic. Journal entries balance
// per commodity, so within one entry the selected-cash postings sum to the
// negation of their counterparts. Each counterpart is then classified on its
// own account class and contributes its own signed amount — there is no
// "first counterpart wins", and a transaction with many counterparts needs no
// allocation rule (reports-plan.md, cashflow rule 5).
func (s *TransactionService) Cashflow(ctx context.Context, input CashflowInput) (CashflowResult, error) {
	startDate, endDate, err := cleanReportDateRange(input.StartDate, input.EndDate)
	if err != nil {
		return CashflowResult{}, err
	}
	bucket := strings.TrimSpace(input.Bucket)
	if !reportBuckets[bucket] {
		return CashflowResult{}, ValidationError{Message: "bucket must be day, week, month, quarter, or year"}
	}

	// The cash scope is resolved once, as of the range end, and applied to the
	// whole range. Re-resolving per bucket would let an account join or leave
	// mid-series, and net_movement would stop reconciling to the balance change
	// of any one stable set of accounts.
	filters, resolvedAccountIDs, err := s.resolveReportFilters(ctx, input.Filters, endDate)
	if err != nil {
		return CashflowResult{}, err
	}

	accounts, err := s.repository.LedgerAccountsAsOf(ctx, BookID, endDate)
	if err != nil {
		return CashflowResult{}, fmt.Errorf("read cashflow accounts: %w", err)
	}

	cashScope := "selected_accounts"
	cashAccountIDs := idSet(resolvedAccountIDs)
	if len(cashAccountIDs) == 0 {
		cashScope = "default_liquid_cash"
		cashAccountIDs = defaultCashAccountIDs(accounts)
		filters.ResolvedAccountIDs = sortedIDs(cashAccountIDs)
	} else if err := validateCashflowScope(accounts, resolvedAccountIDs); err != nil {
		return CashflowResult{}, err
	}

	bounds, err := calendarBucketBounds(startDate, endDate, bucket)
	if err != nil {
		return CashflowResult{}, err
	}

	postings, err := s.repository.ReportCashflowPostings(ctx, db.ReportCashflowPostingsParams{
		BookID:       BookID,
		StartDate:    startDate,
		EndDate:      endDate,
		CommodityIDs: filters.CommodityIDs,
	})
	if err != nil {
		return CashflowResult{}, fmt.Errorf("read cashflow postings: %w", err)
	}

	var rates *RateTable
	if input.Reporting != nil {
		commodityIDs := make([]int64, 0, len(postings))
		for _, posting := range postings {
			commodityIDs = append(commodityIDs, posting.CommodityID)
		}
		rates, err = s.NewRateTable(ctx, *input.Reporting, commodityIDs, startDate, endDate)
		if err != nil {
			return CashflowResult{}, err
		}
	}

	totalsByBucket := make([]*cashflowTotals, len(bounds))
	for index := range totalsByBucket {
		totalsByBucket[index] = newCashflowTotals(rates)
	}

	for _, entry := range groupByJournalEntry(postings) {
		index := bucketIndexFor(bounds, entry.entryDate)
		if index < 0 {
			continue
		}
		classifyCashflowEntry(entry.postings, cashAccountIDs, totalsByBucket[index])
	}

	resultBuckets := make([]CashflowBucket, 0, len(bounds))
	for index, bound := range bounds {
		bucketResult, err := totalsByBucket[index].toBucket(bound.startDate, bound.endDate)
		if err != nil {
			return CashflowResult{}, err
		}
		resultBuckets = append(resultBuckets, bucketResult)
	}

	return CashflowResult{
		StartDate:        startDate,
		EndDate:          endDate,
		Bucket:           bucket,
		Filters:          filters,
		CashScope:        cashScope,
		CashAccountKinds: sortedKinds(defaultCashAccountKinds),
		Buckets:          resultBuckets,
		// System accounts never join the cash scope, but they remain visible as
		// counterparts: transfer clearing is financing movement, not spending.
		ExcludedSystemRoles: []string{"all"},
		Valuation:           s.valuationCoverage(ctx, rates),
	}, nil
}

// classifyCashflowEntry folds one journal entry into a bucket's totals.
//
// A cash-to-cash transfer inside the selected scope disappears here without a
// special case: both legs land in netMovement and cancel, and the entry has no
// counterparts left to classify (cashflow rule 2).
func classifyCashflowEntry(postings []db.ReportCashflowPostingRecord, cashAccountIDs map[int64]bool, totals *cashflowTotals) {
	touchesCash := false
	for _, posting := range postings {
		if cashAccountIDs[posting.AccountID] {
			touchesCash = true
			break
		}
	}
	if !touchesCash {
		return
	}

	for _, posting := range postings {
		if cashAccountIDs[posting.AccountID] {
			addCashflowAmount(totals.netMovement, posting, false)
			continue
		}

		switch posting.AccountClass {
		case "income":
			// Income postings are credits; negating gives a positive inflow.
			addCashflowAmount(totals.inflow, posting, true)
			totals.addConverted(totals.convertedInflow, posting, true)
		case "expense":
			addCashflowAmount(totals.outflow, posting, false)
			totals.addConverted(totals.convertedOutflow, posting, false)
		default:
			// Asset, liability, and equity counterparts outside the selected
			// scope are financing movement. The cash contribution is the
			// negation of the counterpart, so a credit counterpart brings money
			// in and a debit counterpart takes it out.
			if posting.QuantityValue.BigInt().Sign() < 0 {
				addCashflowAmount(totals.transferIn, posting, true)
				totals.addConverted(totals.convertedTransferIn, posting, true)
			} else {
				addCashflowAmount(totals.transferOut, posting, false)
				totals.addConverted(totals.convertedTransferOut, posting, false)
			}
		}
	}
}

func addCashflowAmount(amounts map[int64]*exact.ScaledInt, posting db.ReportCashflowPostingRecord, negate bool) {
	amount := amounts[posting.CommodityID]
	if amount == nil {
		amount = exact.NewScaledInt()
		amounts[posting.CommodityID] = amount
	}
	value := posting.QuantityValue
	if negate {
		value = value.Negated()
	}
	amount.AddCoefficient(value, posting.QuantityScale)
}

func (t *cashflowTotals) toBucket(startDate string, endDate string) (CashflowBucket, error) {
	operatingNet := subtractAmounts(t.inflow, t.outflow)
	transferNet := subtractAmounts(t.transferIn, t.transferOut)

	// Each measure keeps the scale it accumulated at, and each is paired with
	// that scale in the response. Restating them all at the deepest scale any of
	// them reached would read more tidily, but it costs a digit per decimal
	// place against a 38-digit ceiling — so a measure that is representable as
	// recorded could be pushed past the limit by the precision of an unrelated
	// one, turning a valid report into LEDGER_OVERFLOW. Pairing is the client's
	// job and cannot be taken away from it by rewriting the numbers.
	inflow, err := reportMagnitudes(t.inflow, false)
	if err != nil {
		return CashflowBucket{}, err
	}
	outflow, err := reportMagnitudes(t.outflow, false)
	if err != nil {
		return CashflowBucket{}, err
	}
	transferIn, err := reportMagnitudes(t.transferIn, false)
	if err != nil {
		return CashflowBucket{}, err
	}
	transferOut, err := reportMagnitudes(t.transferOut, false)
	if err != nil {
		return CashflowBucket{}, err
	}
	operating, err := reportMagnitudes(operatingNet, false)
	if err != nil {
		return CashflowBucket{}, err
	}
	transfer, err := reportMagnitudes(transferNet, false)
	if err != nil {
		return CashflowBucket{}, err
	}
	netMovement, err := reportMagnitudes(t.netMovement, false)
	if err != nil {
		return CashflowBucket{}, err
	}

	bucket := CashflowBucket{
		StartDate:    startDate,
		EndDate:      endDate,
		Inflow:       inflow,
		Outflow:      outflow,
		TransferIn:   transferIn,
		TransferOut:  transferOut,
		OperatingNet: operating,
		TransferNet:  transfer,
		NetMovement:  netMovement,
	}
	t.attachConverted(&bucket)
	return bucket, nil
}

// attachConverted fills in the reporting-currency measures.
//
// Operating net, transfer net and net movement are computed from the converted
// inflow/outflow/transfer figures rather than accumulated separately, so
// net_movement = operating_net + transfer_net is an identity here exactly as it
// is in the exact figures. Any measure that could not be fully converted stays
// absent, and takes the derived measures that depend on it with it.
func (t *cashflowTotals) attachConverted(bucket *CashflowBucket) {
	if t.rates == nil {
		return
	}

	quantity := func(total *ConvertedTotal) *BalanceQuantity {
		value, scale, ok := total.Value()
		if !ok {
			return nil
		}
		return &BalanceQuantity{
			CommodityID:         t.rates.QuoteCommodityID(),
			QuantityValue:       value,
			QuantityScale:       scale,
			NormalQuantityValue: value,
		}
	}

	inflow := quantity(t.convertedInflow)
	outflow := quantity(t.convertedOutflow)
	transferIn := quantity(t.convertedTransferIn)
	transferOut := quantity(t.convertedTransferOut)

	bucket.ConvertedInflow = inflow
	bucket.ConvertedOutflow = outflow
	bucket.ConvertedTransferIn = transferIn
	bucket.ConvertedTransferOut = transferOut

	operating := differenceOf(inflow, outflow, t.rates)
	transferNet := differenceOf(transferIn, transferOut, t.rates)
	bucket.ConvertedOperatingNet = operating
	bucket.ConvertedTransferNet = transferNet
	bucket.ConvertedNetMovement = sumOf(operating, transferNet, t.rates)
}

// differenceOf and sumOf combine two converted figures at the reporting
// currency's own scale. Both are already rounded there, so this is plain
// integer arithmetic with nothing left to align — and nothing left to round,
// which is what keeps the identity exact.
func differenceOf(left *BalanceQuantity, right *BalanceQuantity, rates *RateTable) *BalanceQuantity {
	if left == nil || right == nil {
		return nil
	}
	value := new(big.Int).Sub(left.QuantityValue.BigInt(), right.QuantityValue.BigInt())
	return convertedQuantity(value, rates)
}

func sumOf(left *BalanceQuantity, right *BalanceQuantity, rates *RateTable) *BalanceQuantity {
	if left == nil || right == nil {
		return nil
	}
	value := new(big.Int).Add(left.QuantityValue.BigInt(), right.QuantityValue.BigInt())
	return convertedQuantity(value, rates)
}

func convertedQuantity(value *big.Int, rates *RateTable) *BalanceQuantity {
	coefficient, err := exact.FromBig(value)
	if err != nil {
		return nil
	}
	return &BalanceQuantity{
		CommodityID:         rates.QuoteCommodityID(),
		QuantityValue:       coefficient,
		QuantityScale:       rates.Scale(),
		NormalQuantityValue: coefficient,
	}
}

// subtractAmounts returns left - right for every commodity present in either,
// keeping the exact scale-aware arithmetic inside ScaledInt.
func subtractAmounts(left map[int64]*exact.ScaledInt, right map[int64]*exact.ScaledInt) map[int64]*exact.ScaledInt {
	result := map[int64]*exact.ScaledInt{}
	for commodityID, amount := range left {
		total := exact.NewScaledInt()
		total.AddScaled(amount)
		result[commodityID] = total
	}
	for commodityID, amount := range right {
		total := result[commodityID]
		if total == nil {
			total = exact.NewScaledInt()
			result[commodityID] = total
		}
		total.SubScaled(amount)
	}
	return result
}

type cashflowEntry struct {
	entryDate string
	postings  []db.ReportCashflowPostingRecord
}

// groupByJournalEntry keeps each entry's postings together, in the order the
// repository returned them.
func groupByJournalEntry(postings []db.ReportCashflowPostingRecord) []cashflowEntry {
	entries := make([]cashflowEntry, 0)
	indexByEntry := map[int64]int{}
	for _, posting := range postings {
		index, ok := indexByEntry[posting.JournalEntryID]
		if !ok {
			entries = append(entries, cashflowEntry{entryDate: posting.EntryDate})
			index = len(entries) - 1
			indexByEntry[posting.JournalEntryID] = index
		}
		entries[index].postings = append(entries[index].postings, posting)
	}
	return entries
}

func bucketIndexFor(bounds []calendarBucketBound, entryDate string) int {
	for index, bound := range bounds {
		if entryDate >= bound.startDate && entryDate <= bound.endDate {
			return index
		}
	}
	return -1
}

// validateCashflowScope adds the one rule this endpoint has beyond the shared
// account-filter contract.
//
// The asset/liability requirement is already enforced for every report by
// validateReportAccountClasses — the cash scope is the balance net_movement
// reconciles to, so a credit card is a legal scope and an income or equity
// account is not. What is specific here is the system-account exclusion: the
// default scope skips them (reports-plan.md, cashflow rule 1) and every
// response reports excluded_system_roles: ["all"], so an explicit selection
// that admitted one would claim to exclude an account it had just summed. That
// rule cannot move into the shared validator, because on net worth
// transfer_clearing is an ordinary basis member.
//
// The resolved selection is checked rather than the named one, because
// include_descendants is what actually becomes the scope.
func validateCashflowScope(accounts []db.LedgerAccountRecord, resolvedAccountIDs []int64) error {
	byID := make(map[int64]db.LedgerAccountRecord, len(accounts))
	for _, account := range accounts {
		byID[account.ID] = account
	}
	for _, accountID := range resolvedAccountIDs {
		account, ok := byID[accountID]
		if !ok {
			// resolveReportFilters already proved the account is in the book. No
			// version as of end_date means it has no postings the range could
			// reach, so it cannot make the scope wrong.
			continue
		}
		if account.SystemRole.Valid {
			return ValidationError{Message: "account filter is invalid: a system account cannot join the cashflow cash scope"}
		}
	}
	return nil
}

// defaultCashAccountIDs is the named liquid-cash default: active, non-system
// accounts of the cash-like kinds.
func defaultCashAccountIDs(accounts []db.LedgerAccountRecord) map[int64]bool {
	ids := map[int64]bool{}
	for _, account := range accounts {
		if account.SystemRole.Valid || account.Status != "active" {
			continue
		}
		if defaultCashAccountKinds[account.AccountKind] {
			ids[account.ID] = true
		}
	}
	return ids
}

func sortedIDs(ids map[int64]bool) []int64 {
	return slices.Sorted(maps.Keys(ids))
}

func sortedKinds(kinds map[string]bool) []string {
	return slices.Sorted(maps.Keys(kinds))
}
